File: chapter3/ex335.hs
import Data.List
import Text.Regex

-- first, a string that matches any number of lower case consonants:
consonants = "[bcdfghjlkmnpqrstvwxyz]*"
vowels = ["a+", "e+", "i+", "o+", "u+"]

vowelsRegex = mkRegex $ intercalate consonants vowels

-- 3.3.5 b)
lexOrderStr = [ a:"*" | a <- ['a'..'z'] ]
lexOrderRegex = (mkRegex . concat) lexOrderStr

File: chapter3/ex_3_3_5.py
# Exercise 3.3.5:
# Write regular definitions for the following languages:

import re 
import string

# a) All strings of lowercase letters that contain the five vowels in 
# order.

# In other words, any number (including 0) of consonants, then a, 
# then any number of consonants, then e, and so on and so forth...
consonants = 'bcdfghjlkmnpqrstvwxyz'

lowercase_re = re.compile('[{c}]*a+[{c}]*e+[{c}]*i+[{c}]*o+[{c}]*u+[{c}]'.format(
                            c=consonants))

# b) All strings of lowercase letters in which the letters are in 
# ascending lexicographic order.

# In other words: any number of as, followed by any number of bs, 
# so on and so forth, until we reach z.
# More formally, the regular def. for this language would be 
# something like:
# a_def -> 'a*'
# b_def -> a_def|'b*'
# c_def -> b|c*
# etc.
# Sigma is {}.
lex_asc_re = re.compile('a*b*c*d*e*f*g*h*i*j*k*l*m*n*o*p*q*r*s*t*u*v*w*x*y*z*')
